client/session.go

```diff
@@ -574,99 +651,68 @@ func (s *session) Write(namespace, id string, t time.Time, value float64, unit x
-				wg.Done()
-			}
-		case topology.ConsistencyLevelAll:
-			if doneAll && atomic.CompareAndSwapInt32(&wgIsDone, 0, 1) {
-				wg.Done()
-			}
-		}
-		// Finalize resources that can be returned
-		if doneAll {
-			// Return write to pool
-			s.writeOpPool.Put(w)
-		}
-	}
+	c.op, c.majority = s.writeOpPool.Get(), majority
 
-	routeErr := s.topoMap.RouteForEach(tsID, func(idx int, host topology.Host) {
+	c.op.namespace = ts.StringID(namespace)
+	c.op.request.ID = tsID.Data()
+	c.op.request.Datapoint.Value = value
+	c.op.request.Datapoint.Timestamp = timestamp
+	c.op.request.Datapoint.TimestampType = timeType
+	c.op.request.Datapoint.Annotation = annotation
+	c.op.completionFn = c.completionFn
+
+	if err := s.topoMap.RouteForEach(tsID, func(idx int, host topology.Host) {
 		// First count all the pending write requests to ensure
 		// we count the amount we're going to be waiting for
 		// before we enqueue the completion fns that rely on having
 		// an accurate number of the pending requests when they execute
-		pending++
-	})
-	if routeErr == nil {
-		// Now enqueue the write requests
-		routeErr = s.topoMap.RouteForEach(tsID, func(idx int, host topology.Host) {
-			enqueued++
-			if err := s.queues[idx].Enqueue(w); err != nil && enqueueErr == nil {
-				// NB(r): if this ever happens we have a code bug, once we
-				// are in the read lock the current queues we are using should
-				// never be closed
-				enqueueErr = err
-			}
-		})
+		c.pending++
+		c.queues = append(c.queues, s.queues[idx])
+	}); err != nil {
+		c.decref()
+		s.RUnlock()
+		return err
 	}
-	s.RUnlock()
 
-	if routeErr != nil {
-		return routeErr
-	}
+	c.Lock()
 
-	if enqueueErr != nil {
-		s.opts.InstrumentOptions().Logger().Errorf("failed to enqueue write: %v", enqueueErr)
-		return enqueueErr
-	}
+	for i := range c.queues {
+		if err := c.queues[i].Enqueue(c.op); err != nil {
+			c.Unlock()
+			c.decref()
 
-	// Wait for writes to complete
-	wg.Wait()
+			// NB(r): if this ever happens we have a code bug, once we
+			// are in the read lock the current queues we are using should
+			// never be closed
+			s.RUnlock()
+			s.opts.InstrumentOptions().Logger().Errorf("failed to enqueue write: %v", err)
+			return err
+		}
 
-	var reportErrors []error
-	errsLen := atomic.LoadInt32(&resultErrs)
-	if errsLen > 0 {
-		resultErrLock.RLock()
-		reportErrors = errors[:]
-		resultErrLock.RUnlock()
+		c.incref()
+		enqueued++
 	}
-	responded := enqueued - atomic.LoadInt32(&pending)
-	err := s.writeConsistencyResult(majority, enqueued, responded, errsLen, reportErrors)
-	s.incWriteMetrics(err, errsLen)
+
+	// Wait for writes to complete. We don't need to loop over Wait() since there
+	// are no spurious wakeups in Golang and the condition is one-way and binary.
+	s.RUnlock()
+	c.condition.Wait()
+
+	err := s.writeConsistencyResult(majority,
+		enqueued, enqueued-atomic.LoadInt32(&c.pending), int32(len(c.errors)), c.errors)
+	s.incWriteMetrics(err, int32(len(c.errors)))
```

Comment:
Maybe we should add two methods to `writeState` - namely `Errors() []error` and `Pending() int32` that encapsulate perform atomic.Load, etc better separation of concerns?  i.e. I don't need to know how `writeState` is implemented in this `Write(...)` call if these are encapsulated, so less cognitive load when reading this method.